arch/riscv/core/pmp.c

```diff
@@ -534,51 +628,63 @@ static inline unsigned int z_riscv_pmp_thread_init(unsigned long *pmp_addr,
 	 * Retrieve the pmpaddr value matching the last global PMP slot.
 	 * This is so that set_pmp_entry() can safely attempt TOR with it.
 	 */
-	pmp_addr[global_pmp_end_index - 1] = global_pmp_last_addr;
+#ifdef CONFIG_USERSPACE
+	pmp_addr[global_pmp_u_end_index - 1] = global_pmp_u_last_addr;
+	return global_pmp_u_end_index;
+#else
+	pmp_addr[global_pmp_m_end_index - 1] = global_pmp_m_last_addr;
+	return global_pmp_m_end_index;
+#endif /*CONFIG_USERSPACE */
```

Comment:
When CONFIG_USERSPACE=y, `z_riscv_pmp_kernelmode_prepare` should not get `global_pmp_u_last_addr` and `global_pmp_u_end_index`.